File: 06__full-stack/brainrot-studio/backend/app/schemas/wizard.py
from uuid import UUID
from pydantic import BaseModel, Field


class CharacterAssetRead(BaseModel):
    id: UUID
    character_id: str
    asset_type: str
    expression: str
    image_url: str


class CharacterRead(BaseModel):
    id: str
    name: str
    show_name: str | None = None
    image_url: str
    avatar: str
    tier: str
    assets: list[CharacterAssetRead] = []


class NicheRead(BaseModel):
    id: UUID
    name: str
    slug: str
    description: str | None = None
    icon: str = "🎬"


class GeneratedTopicSchema(BaseModel):
    id: UUID | None = None
    title: str = Field(description="Catchy title for the Short")
    hook: str = Field(description="Initial hook or opening premise")
    premise: str = Field(description="Core narrative summary")
    estimated_duration: int = Field(default=35, description="Duration in seconds")
    characters: list[str] = Field(default=[], description="List of character IDs involved")


class DialogueLineSchema(BaseModel):
    character_id: str = Field(description="Character ID or name speaking")
    text: str = Field(description="Dialogue line spoken")


class ScriptSceneSchema(BaseModel):
    scene_number: int
    visual_description: str = Field(description="Detailed visual direction for canvas background")
    dialogue: list[DialogueLineSchema] = Field(default=[])
    duration_seconds: float = Field(default=6.0)


class GeneratedScriptSchema(BaseModel):
    id: UUID | None = None
    title: str
    hook: str
    scenes: list[ScriptSceneSchema]
    estimated_duration: float = Field(default=35.0)


class VideoStyleConfigSchema(BaseModel):
    layout: str = Field(default="centered", description="Layout preset e.g. centered, classic, top_caption, character_dialogue")
    font_family: str = Field(default="Inter", description="Font family name e.g. Inter, Impact, Outfit")
    font_size: int = Field(default=72, description="Font size in px")
    primary_color: str = Field(default="#FFFFFF", description="Hex text color")
    outline_color: str = Field(default="#000000", description="Hex outline color")
    outline_width: int = Field(default=4, description="Outline width in px")
    animation: str = Field(default="pop", description="Subtitle animation style e.g. pop, fade, bounce")
    position: str = Field(default="center", description="Position on screen")


class CreateSessionRequest(BaseModel):
    project_id: UUID | None = None


class UpdateSessionRequest(BaseModel):
    current_step: int | None = None
    character_ids: list[str] | None = None
    niche: str | None = None


class SelectTopicRequest(BaseModel):
    topic_id: UUID


class RegenerateSceneRequest(BaseModel):
    instruction: str = Field(description="Instruction on how to regenerate this specific scene")


class GenerateStyleRequest(BaseModel):
    prompt: str = Field(description="Natural language prompt describing the desired video style")

File: 06__full-stack/brainrot-studio/frontend/src/lib/api/wizard.ts
import { api } from './client';

export interface CharacterAsset {
	id: string;
	character_id: string;
	asset_type: string;
	expression: string;
	image_url: string;
}

export interface Character {
	id: string;
	name: string;
	show_name?: string;
	image_url: string;
	avatar: string;
	tier: string;
	selected?: boolean;
	assets?: CharacterAsset[];
}

export interface Niche {
	id: string;
	name: string;
	slug: string;
	description?: string;
	icon: string;
}

export interface GeneratedTopic {
	id: string;
	title: string;
	hook: string;
	premise: string;
	estimated_duration: number;
	characters: string[];
}

export interface DialogueLine {
	character_id: string;
	text: string;
}

export interface ScriptScene {
	scene_number: number;
	visual_description: string;
	dialogue: DialogueLine[];
	duration_seconds: number;
}

export interface GeneratedScript {
	id?: string;
	title: string;
	hook: string;
	scenes: ScriptScene[];
	estimated_duration: number;
}

export interface VideoStyleConfig {
	layout: string;
	font_family: string;
	font_size: number;
	primary_color: string;
	outline_color: string;
	outline_width: number;
	animation: string;
	position: string;
}

export interface GenerationSession {
	id: string;
	status: string;
	current_step: number;
	niche?: string;
	character_ids: string[];
	selected_topic_id?: string;
	topics: GeneratedTopic[];
	script?: GeneratedScript;
	style_config?: VideoStyleConfig;
}

export async function fetchCharacters(): Promise<Character[]> {
	return api<Character[]>('/characters');
}

export async function fetchNiches(): Promise<Niche[]> {
	return api<Niche[]>('/niches');
}

export async function createGenerationSession(projectId?: string): Promise<{ id: string; status: string; current_step: number }> {
	return api('/generation-sessions', {
		method: 'POST',
		body: JSON.stringify({ project_id: projectId })
	});
}

export async function getGenerationSession(sessionId: string): Promise<GenerationSession> {
	return api<GenerationSession>(`/generation-sessions/${sessionId}`);
}

export async function updateGenerationSession(
	sessionId: string,
	data: { current_step?: number; character_ids?: string[]; niche?: string }
): Promise<{ status: string }> {
	return api(`/generation-sessions/${sessionId}`, {
		method: 'PATCH',
		body: JSON.stringify(data)
	});
}

export async function generateSessionTopics(sessionId: string): Promise<GeneratedTopic[]> {
	return api<GeneratedTopic[]>(`/generation-sessions/${sessionId}/topics`, {
		method: 'POST'
	});
}

export async function selectSessionTopic(sessionId: string, topicId: string): Promise<{ status: string }> {
	return api(`/generation-sessions/${sessionId}/select-topic`, {
		method: 'POST',
		body: JSON.stringify({ topic_id: topicId })
	});
}

export async function generateSessionScript(sessionId: string): Promise<GeneratedScript> {
	return api<GeneratedScript>(`/generation-sessions/${sessionId}/script`, {
		method: 'POST'
	});
}

export async function regenerateSessionScene(
	sessionId: string,
	sceneIdx: number,
	instruction: string
): Promise<ScriptScene> {
	return api<ScriptScene>(`/generation-sessions/${sessionId}/script/scenes/${sceneIdx}`, {
		method: 'PATCH',
		body: JSON.stringify({ instruction })
	});
}

export async function generateSessionStyle(sessionId: string, prompt: string): Promise<VideoStyleConfig> {
	return api<VideoStyleConfig>(`/generation-sessions/${sessionId}/style`, {
		method: 'POST',
		body: JSON.stringify({ prompt })
	});
}

export async function renderSessionVideo(sessionId: string): Promise<{ status: string; video_id: string; project_id: string }> {
	return api(`/generation-sessions/${sessionId}/render`, {
		method: 'POST'
	});
}
